### ISSUE TITLE
assigning 'this' to option-type result causes interpreter crash

### ISSUE THREAD

**Issue by fridis** [tokiwa-software/fuzion#141]
For a value type 'B', returning 'B.this' in an inner feature with result 'option<B>' causes the interpreter to crash:
```
 > cat matchThis.fz
matchThis is

  A ref is
    getA option<A> is A.this

  match A.getA
    a A => say "got A"
    nil => say "got nil"

  B is
    getB option<B> is B.this

  match B.getB
    b B => say "got B"
    nil => say "got nil"
 > ./build/bin/fz matchThis.fz 
got A

error 1: *** java.lang.ArrayIndexOutOfBoundsException: Index 0 out of bounds for length 0
Call stack:
matchThis: /home/fridi/fuzion/work/matchThis.fz:13:11:
  match B.getB
----------^


java.lang.ArrayIndexOutOfBoundsException: Index 0 out of bounds for length 0
	at dev.flang.be.interpreter.Instance.storeNonRef(Instance.java:387)
	at dev.flang.be.interpreter.Instance.storeNonRef(Instance.java:367)
	at dev.flang.be.interpreter.Interpreter.setNonRefField(Interpreter.java:1116)
	at dev.flang.be.interpreter.Interpreter.setFieldSlot(Interpreter.java:1319)
	at dev.flang.be.interpreter.Interpreter.setField(Interpreter.java:1348)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:339)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:372)
	at dev.flang.be.interpreter.Interpreter.callOnInstance(Interpreter.java:896)
	at dev.flang.be.interpreter.Interpreter.lambda$callable$6(Interpreter.java:791)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:325)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:404)
	at dev.flang.be.interpreter.Interpreter.execute(Interpreter.java:372)
	at dev.flang.be.interpreter.Interpreter.callOnInstance(Interpreter.java:896)
	at dev.flang.be.interpreter.Interpreter.lambda$callable$6(Interpreter.java:791)
	at dev.flang.be.interpreter.Interpreter.run(Interpreter.java:235)
	at dev.flang.tools.Fuzion.lambda$parseArgsForBackend$3(Fuzion.java:549)
	at dev.flang.tools.Tool.run(Tool.java:168)
	at dev.flang.tools.Fuzion.main(Fuzion.java:244)

```
The C backend does not like this either:
```
 > ./build/bin/fz -c matchThis.fz 
java.util.EmptyStackException
	at java.base/java.util.Stack.peek(Stack.java:101)
	at java.base/java.util.Stack.pop(Stack.java:83)
	at dev.flang.be.c.C.pop(C.java:260)
	at dev.flang.be.c.C.assignField(C.java:730)
	at dev.flang.be.c.C.access(C.java:408)
	at dev.flang.be.c.C.createCode(C.java:460)
	at dev.flang.be.c.C.createCode(C.java:293)
	at dev.flang.be.c.C.codeForRoutine(C.java:1012)
	at dev.flang.be.c.C.code(C.java:956)
	at dev.flang.be.c.C$CompilePhase$4.compile(C.java:67)
	at dev.flang.be.c.C.lambda$createCode$0(C.java:207)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)
	at java.base/java.util.stream.ReferencePipeline$Head.forEachOrdered(ReferencePipeline.java:772)
	at dev.flang.be.c.C.createCode(C.java:203)
	at dev.flang.be.c.C.compile(C.java:152)
	at dev.flang.tools.Fuzion.lambda$parseArgsForBackend$3(Fuzion.java:551)
	at dev.flang.tools.Tool.run(Tool.java:168)
	at dev.flang.tools.Fuzion.main(Fuzion.java:244)
```

**Comment by fridis** [comment]
After changing the syntax to match the current language:
```
matchThis is

  A ref is
    getA option A is A.this

  match A.getA
    a A => say "got A"
    nil => say "got nil"

  B is
    getB option B is B.this

  match B.getB
    b B => say "got B"
    nil => say "got nil"
```
The example produces a correct error since `B.this` is not compatible to `B`
```
 > ./build/bin/fz ex_141.fz

/home/fridi/fuzion/clean/fuzion.1/ex_141.fz:11:24: error 1: Incompatible types in assignment
    getB option B is B.this
-----------------------^
assignment to field : 'matchThis.B.getB.result'
expected formal type: 'option matchThis.this.type.B'
actual type found   : 'matchThis.this.type.B.this.type'
assignable to       : ref 'matchThis.this.type.ref B'
for value assigned  : 
{
  deref(getB.this.#^matchThis.B.getB)
}

To solve this, you could change the type of the target 'matchThis.B.getB.result' to 'matchThis.this.type.B.this.type' or convert the type of the assigned value to 'option matchThis.this.type.B'.

one error.
```
Changing the result type for `getB` to `option B.this.type` solves this:
```
matchThis is

  A ref is
    getA option A is A.this

  match A.getA
    a A => say "got A"
    nil => say "got nil"

  B is
    getB option B.this.type is B.this

  match B.getB
    b B => say "got B"
    nil => say "got nil"
```
```
 > PRECONDITIONS=true POSTCONDITIONS=true ./build/bin/fz ex_141c.fz
got A
got B
```
However, an attempt to fix this using `fixed` causes a crash (both in interpreter and C backend):
```
 > cat ex_141b.fz
matchThis is

  A ref is
    getA option A is A.this

  match A.getA
    a A => say "got A"
    nil => say "got nil"

  B is
    fixed getB option B is B.this

  match B.getB
    b B => say "got B"
    nil => say "got nil"
```
```
> PRECONDITIONS=true POSTCONDITIONS=true ./build/bin/fz -c ex_141b.fz

error 1: java.lang.Error: require-condition1 failed: dev.flang.fuir.analysis.AbstractInterpreter:pop:311
	at dev.flang.util.ANY.require(ANY.java:93)
	at dev.flang.fuir.analysis.AbstractInterpreter.pop(AbstractInterpreter.java:311)
	at dev.flang.fuir.analysis.AbstractInterpreter.process(AbstractInterpreter.java:529)
	at dev.flang.fuir.analysis.AbstractInterpreter.process(AbstractInterpreter.java:446)
	at dev.flang.fuir.analysis.AbstractInterpreter.process(AbstractInterpreter.java:412)
	at dev.flang.fuir.analysis.dfa.DFA.analyze(DFA.java:882)
	at dev.flang.fuir.analysis.dfa.DFA.iteration(DFA.java:847)
	at dev.flang.fuir.analysis.dfa.DFA.findFixPoint(DFA.java:810)
	at dev.flang.fuir.analysis.dfa.DFA.dfa(DFA.java:792)
	at dev.flang.fuir.analysis.dfa.DFA.new_fuir(DFA.java:721)
	at dev.flang.be.c.C.<init>(C.java:495)
	at dev.flang.tools.Fuzion$Backend$2.process(Fuzion.java:145)
	at dev.flang.tools.Fuzion$Backend.processFrontEnd(Fuzion.java:368)
	at dev.flang.tools.Fuzion.lambda$parseArgsForBackend$3(Fuzion.java:832)
	at dev.flang.tools.Tool.lambda$run$0(Tool.java:154)
	at dev.flang.util.Errors.runAndExit(Errors.java:704)
	at dev.flang.tools.Tool.run(Tool.java:154)
	at dev.flang.tools.Fuzion.main(Fuzion.java:481)


*** fatal errors encountered, stopping.
```
Will create a new issue for that.

**Comment by fridis** [comment]
The original issue is fixed with proper support for outer's this type (#1209), the remaining issue with `fixed` has been moved to a new issue (#1214).
